libvast/src/data.cpp

```diff
@@ -285,4 +338,108 @@ bool convert(const data& d, json& j, const type& t) {
   return convert(d, j);
 }
 
+namespace {
+
+data parse(const YAML::Node& node) {
+  switch (node.Type()) {
+    case YAML::NodeType::Undefined:
+    case YAML::NodeType::Null:
+      return data{};
+    case YAML::NodeType::Scalar: {
+      auto str = node.as<std::string>();
+      data result;
+      // Attempt some type inference.
+      if (parsers::boolean(str, result))
+        return result;
+      // Attempt maximum type inference.
+      if (parsers::data(str, result))
+        return result;
+      // Take the input as-is if nothing worked.
+      return str;
+    }
+    case YAML::NodeType::Sequence: {
+      list xs;
+      xs.reserve(node.size());
+      for (size_t i = 0; i < node.size(); ++i)
+        xs.push_back(parse(node[i]));
+      return xs;
+    }
+    case YAML::NodeType::Map: {
+      record xs;
+      xs.reserve(node.size());
+      for (auto& pair : node)
+        xs.emplace(pair.first.as<std::string>(), parse(pair.second));
+      return xs;
+    }
+  }
+  return {}; // makes GCC happy
+}
+
+} // namespace
+
+caf::expected<data> from_yaml(std::string_view str) {
+  try {
+    // Maybe one glory day in the future it will be possible to perform a
+    // single pass over the input without incurring a copy.
+    auto node = YAML::Load(std::string{str});
+    return parse(node);
+  } catch (const YAML::Exception& e) {
+    return caf::make_error(ec::parse_error,
+                           "failed to parse YAML at line/col/pos", e.mark.line,
+                           e.mark.column, e.mark.pos);
+  }
+}
+
+namespace {
+
+void print(YAML::Emitter& out, const data& x) {
+  // clang-format off
+  auto f = detail::overload(
+    [&out](const auto& x) { out << to_string(x); },
+    [&out](bool x) { out << (x ? "true" : "false"); },
+    [&out](count x) { out << x; },
+    [&out](integer x) { out << x; },
+    [&out](real x) { out << x; },
+    [&out](const std::string& x) { out << x; },
+    [&out](const list& xs) {
+      out << YAML::BeginSeq;
+      for (auto& x : xs)
+        print(out, x);
+      out << YAML::EndSeq;
+    },
+    // We treat maps like records.
+    [&out](const map& xs) {
+      out << YAML::BeginMap;
+      for (auto& [k, v] : xs) {
+        out << YAML::Key;
+        print(out, k);
+        out << YAML::Value;
+        print(out, v);
+      }
+      out << YAML::EndMap;
+    },
+    [&out](const record& xs) {
+      out << YAML::BeginMap;
+      for (auto& [k, v] : xs) {
+        out << YAML::Key << k << YAML::Value;
+        print(out, v);
+      }
+      out << YAML::EndMap;
```

Comment:
It'd be more idiomatic to append `YAML::EndMap` in a scope guard that you set up immediately after appending `YAML::BeginMap`. The same applies to all other places.

Reply:
While safer, it is still more complicated to read. If you start with the tutorial of the official package, you can immediately recognize the patttern.

If the control flow would be more convoluted, I'd fully agree, but adding RAII to these few lines are not adding much value IMO.

Reply:
RAII isn't a matter of readability, but rather of security. yaml-cpp uses exceptions, and if the code between the begin and end statements throws, we still want the end statement to be executed.

Reply:
@lava mind being the decider here, again?

Reply:
Several issues with this:
 - If yaml-cpp indeed uses exceptions on the print path, the `to_yaml()` function below needs a try-catch block as well
 - If printing can throw, we can't blindly print in a scope guard since that might throw again. Instead we'd have to check `std::active_exceptions()`.
 - Printing `EndMap` is part of the happy path only. There is no guarantee that `Yaml::EndMap` is syntactically correct when an exception is thrown (i.e. if the exception happens between printing a key and a value). So we have to discard the result anyways after an exception.
- We don't leak resources by not printing `YAML::EndMap`

So, I'm in favor of leaving the code as-is. The first bullet point should still be addressed, though.